server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1366,11 +1366,12 @@ public RefreshStats refreshStats() {
     }
 
     public FlushStats flushStats() {
+        final Engine engine = getEngineOrNull();
         return new FlushStats(
             flushMetric.count(),
             periodicFlushMetric.count(),
             TimeUnit.NANOSECONDS.toMillis(flushMetric.sum()),
-            getEngineOrNull() != null ? getEngineOrNull().getTotalFlushTimeExcludingWaitingOnLockInMillis() : 0L
+            engine != null ? engine.getTotalFlushTimeExcludingWaitingOnLockInMillis() : 0L
```

Comment:
nit Note we cannot use withEngineOrNull because the method can block and cannot be called from cluster state update threads.

Reply:
don't we have the same issue with `getEngineOrNull`?

Reply:
Hi @fcofdez , no, the fact we got the NPE when calling getTotalFlushTimeExcludingWaitingOnLockInMillis with the previous version of the code was that the first invocation of getEngineOrNull() returned a non-null engine and then proceeded to call getTotalFlushTimeExcludingWaitingOnLockInMillis() on getEngineOrNull() again which now returned null. So the engine was closed in between the two getEngineOrNull() calls. Now, we get the engine just once at the beginning of the function, so it's either null, and we return 0, or it's not null, and we get the getTotalFlushTimeExcludingWaitingOnLockInMillis (either from an open or closed engine, but an existing reference nonetheless, so no NPE).

Reply:
Oh @fcofdez just realized your question.

assertCurrentThreadWithEngine() is only in the withEngine() methods. We've not made it in the general getEngineOrNull() method (maybe because I suspect we'd have a lot of assertions failing). However, indeed we've noticed commitStats() being problematic around super thin shards in https://elasticco.atlassian.net/browse/ES-12215, so I'm extending that point on improving how the general IndexShard stats method might be improved to be accessed around the super thin project.

At this point, we should not be concerned about super thin for this core ES PR I believe.